torch/fx/passes/runtime_assert.py

```diff
@@ -85,60 +93,140 @@ def insert_deferred_runtime_asserts(
     for ras in ras_by_symbol.values():
         for ra in ras:
             needed_symbols.update(free_symbols(ra.expr))
+    for node in gm.graph.nodes:
+        if (
+            node.op == "call_function"
+            and _is_intermediate_tensor_shape_call(node)
+        ):  # use symbols for intermediate shape calls, even if not in runtime asserts
+            sym_expr = _get_sym_val(node)
+            needed_symbols.update(sym_expr.free_symbols)
 
     log.debug("needed_symbols = %s", needed_symbols)
 
+    # Track asserts/checks we've added
+    added_asserts: Set["sympy.Expr"] = set()
+    constrained_unbacked_symbols: Set["sympy.Symbol"] = set()
+
+    def _sympy_interp(symbol_to_proxy, expr):
+        # sympy_interp() with hash consing
+        from sympy import Integer, Number, Symbol
+        from sympy.logic.boolalg import BooleanAtom
+        from torch.utils._sympy.interp import _run_sympy_handler, sympy_interp
+
+        # base cases, don't cache
+        if isinstance(expr, (Integer, Number, Symbol, BooleanAtom)):
+            return sympy_interp(
+                PythonReferenceAnalysis, symbol_to_proxy, expr
+            )
+        # hash cons
+        if expr in symbol_to_proxy:
+            return symbol_to_proxy[expr]
+
+        # hash cons on arguments
+        fx_args = [
+            (
+                symbol_to_proxy[arg]
+                if arg in symbol_to_proxy
+                else _sympy_interp(symbol_to_proxy, arg)
+            )
+            for arg in expr.args
+        ]
+
+        # run expr handler
+        symbol_to_proxy[expr] = _run_sympy_handler(
+            PythonReferenceAnalysis, fx_args, expr
+        )
+        return symbol_to_proxy[expr]
+
+    def _is_bound_expr_for_symbol(expr: "sympy.Expr") -> bool:
+        # This is probably unnecessary, but since torch._check() calls for single-symbol bounds
+        # like u0 >= 0, 10 >= u0 accumulate range info in the ShapeEnv, and we insert sym_constrain_range calls
+        # anyways, we designate these calls as redundant and remove them.
+        if (
+            len(expr.args) != 2
+            or expr.func not in (sympy.LessThan, sympy.GreaterThan)
+        ):
+            return False
+        lhs, rhs = expr.args
+        return (
+            (isinstance(lhs, sympy.Symbol) and isinstance(rhs, sympy.Number))
+            or (isinstance(rhs, sympy.Symbol) and isinstance(lhs, sympy.Number))
+        )
+
     def add_runtime_asserts(ras):
         for ra in ras:
+            if (
+                ra.expr in added_asserts  # redundant
+                or (
+                    len(ra.expr.free_symbols) == 1
+                    and next(iter(ra.expr.free_symbols)) in constrained_unbacked_symbols
+                    and _is_bound_expr_for_symbol(ra.expr)
+                )
+                # if we've already added a constrain_range call for this symbol,
+                # then single-symbol bound asserts like u0 >= 0, u0 <= 5 are redundant.
```

Comment:
I think my preference is that constrain ranges always turn into a single symbol assert like u0 >= 0, then everything is covered by `added_asserts`